sklearn/preprocessing/data.py

```diff
@@ -1660,14 +1673,24 @@ def _transform_selected(X, transform, selected="all", copy=True):
     elif n_selected == n_features:
         # All features selected.
         return transform(X)
-    else:
-        X_sel = transform(X[:, ind[sel]])
-        X_not_sel = X[:, ind[not_sel]]
 
-        if sparse.issparse(X_sel) or sparse.issparse(X_not_sel):
-            return sparse.hstack((X_sel, X_not_sel))
-        else:
-            return np.hstack((X_sel, X_not_sel))
+    X_sel = transform(X[:, ind[sel]])
+    X_not_sel = X[:, ind[not_sel]]
+
+    if retain_order:
+        if X_sel.shape[1] + X_not_sel.shape[1] != n_features:
+            raise ValueError("The retain_order option can only be set to True "
+                             "if the dimensions of the input array match the "
+                             "dimensions of the transformed array.")
+
+        # Fancy indexing not supported for sparse matrices
```

Comment:
FWIW it's not particularly hard to do this kind of splicing operation. Three options:
* create a sparse diagonal matrix `D` of size `(X_sel.shape[1], X.shape[1])` such that `X_sel @ D` effectively inserts empty columns in `X_sel`. Then do the same for `X_not_sel` and sum them together. (Easiest to read?)
* convert to COO, map `.col` to the new space in both `X_sel` and `X_not_sel`, then create a new COO from the concatenation of their `row`, `col` and `data`.
* convert to CSC, and use `np.insert` on `indices`, on `data` and on `np.diff(indptr)` to define the output matrix. (Most efficient but hardest to implement.)

But it's fine being crippled for now.